File: Frontend/failed_in/lib/models/notification_model.dart
enum AppNotificationType {
  Comment,
  Like,
  Report,
  Reply,
}

class AppNotification {
  String? id;
  String? senderId;
  String? receiverId;
  String? postId;
  String? message;
  bool? viewed;
  String? postTitle;
  String? type;
  String? time;

  AppNotification({
    this.id,
    this.senderId,
    this.receiverId,
    this.postId,
    this.postTitle,
    this.message,
    this.viewed,
    this.type,
    this.time,
  });

  static AppNotification fromJson(data) {
    return AppNotification(
      id: data['_id'],
      message: data['message'],
      postId: data['postId']['_id'],
      postTitle: data['postId']['title'],
      senderId: data['senderId']['username'],
      receiverId: data['receiverId']['_id'],
      viewed: data['viewed'],
      type: 'AppNotificationType.' + data['type'],
      time: data['createdAt'],
    );
  }
}

File: Backend/models/notificationModel.js
const mongoose = require('mongoose');

const notificationSchema = mongoose.Schema({
    type: {
        type: String,
        enum: ['Comment', 'Like', 'Report', 'Reply'],
        required: true,
    },
    senderId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
    },
    receiverId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    postId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Post',
        required: true
    },
    message: {
        type: String
    },
    viewed: {
        type: Boolean,
        default: false
    },  
    expire_at: {
        type: Date
    } 
}, { timestamps: true }).index({"expire_at": 1 }, { expireAfterSeconds: 30 } );;

//populating the senderId and postID in notificationSchema
notificationSchema.pre('find', async function(next) {
    this.populate('senderId', ['username', 'image']);
    this.populate('postId', ['title']);
    next();
});

const Notification = mongoose.model('Notification', notificationSchema);

module.exports = Notification;